airsonic-main/src/main/java/org/airsonic/player/service/SecurityService.java

```diff
@@ -347,15 +395,111 @@ protected boolean isFileInFolder(String file, String folder) {
+        if (saved != null && saved.identical(sonosLink)) {
+            setUser(sonosLink.getUsername());
+        } else {
+            throw new SonosSoapFault.LoginUnauthorized();
+        }
     }
 
-    public void setUserCache(Ehcache userCache) {
-        this.userCache = userCache;
+    public void authenticate() throws SonosSoapFault.LoginUnauthorized {
+        setUser(USERNAME_SONOS);
     }
-}
\ No newline at end of file
+
+
+    /**
+     * Generate a link code, and put in the cache for future use.
+     *
+     * @param householdId from sonos, represent a user on sonos
+
+     * @return The link code.
+     */
+    public String generateLinkCode(String householdId) {
+        String linkCode = createLinkCode();
+
+        sonosLinkcodeCache.put(new Element(linkCode, householdId));
+
+        return linkCode;
+    }
+
+    public String getHousehold(String linkCode) {
+        Element element = sonosLinkcodeCache.get(linkCode);
+
+        if (element != null) {
+            return (String) element.getValue();
+        }
+
+        return null;
+    }
+
+    /**
+     * Insert the authorisation in sonoslink. Verify is the linkcode exist, and cannot be get again.
+     *
+     * @param username The username authorisation for sonos link
+     * @param householdId The id of entry in sonos controller
+     * @param linkcode The link code used between sonos and airsonic.
+     * @return true if the insert is ok, false if some entry exist with the linkcode
+     */
+    public boolean authoriseSonos(String username, String householdId, String linkcode) {
+        if (sonosLinkDao.findByLinkcode(linkcode) != null) {
+            return false;
+        }
+
+        sonosLinkDao.create(new SonosLink(username, householdId, linkcode));
+        return true;
+    }
+
+    /**
+     * Find the user they have a linkCode for the householdId set, build the authToken and return it.
+     *
+     * @param householdId The householdId from Sonos
+     * @param linkCode The linkCode return it before
+     * @return The build authToken or null if didn't find any user with householdId and linkCode
+     */
+    public String getSonosAuthToken(String householdId, String linkCode) {
+
+        SonosLink sonosLink = sonosLinkDao.findByLinkcode(linkCode);
+        if (sonosLink != null && householdId.equals(sonosLink.getHouseholdid())) {
+            return buildToken(sonosLink);
+        } else {
+            return null;
+        }
+    }
+
+    private String buildToken(SonosLink link) {
+        return jwtSecurityService.createSonosToken(link.getUsername(), link.getHouseholdid(), link.getLinkcode());
+    }
+
+    public SonosLink getSonosLink(String linkCode) {
+        return sonosLinkDao.findByLinkcode(linkCode);
+    }
+
+    private void setUser(String username) throws SonosSoapFault.LoginUnauthorized {
+        User user = getUserByName(username, true);
+        Authentication authentication = authenticate(user.getUsername(), user.getPassword());
```

Comment:
do you need to auth here?
as in what good does fetching a password from the db and using the same password to authenticate against the same db do here? this will always be a positive match, no?